mm2src/coins/eth.rs

```diff
@@ -1243,55 +1245,348 @@ impl SwapOps for EthCoin {
+
+        let selfi = self.clone();
+        let swap_id = selfi.etomic_swap_id(input.time_lock, &input.secret_hash);
+        let secret_hash = if input.secret_hash.len() == 32 {
+            ripemd160(&input.secret_hash).to_vec()
+        } else {
+            input.secret_hash.to_vec()
+        };
+        let fut = async move {
+            let status = selfi
+                .payment_status(expected_swap_contract_address, Token::FixedBytes(swap_id.clone()))
+                .compat()
+                .await
+                .map_to_mm(ValidatePaymentError::Transport)?;
+            if status != PAYMENT_STATE_SENT.into() {
+                return MmError::err(ValidatePaymentError::UnexpectedPaymentState(format!(
+                    "Payment state is not PAYMENT_STATE_SENT, got {}",
+                    status
+                )));
+            }
+
+            let tx_from_rpc = selfi
+                .web3
+                .eth()
+                .transaction(TransactionId::Hash(tx.hash))
+                .compat()
+                .await?;
+            let tx_from_rpc = match tx_from_rpc {
+                Some(t) => t,
+                None => {
+                    return MmError::err(ValidatePaymentError::InvalidRpcResponse(format!(
+                        "Didn't find provided tx {:?} on ETH node",
+                        tx
+                    )))
+                },
+            };
+
+            if tx_from_rpc.from != sender {
+                return MmError::err(ValidatePaymentError::WrongPaymentTx(format!(
+                    "Payment tx {:?} was sent from wrong address, expected {:?}",
+                    tx_from_rpc, sender
+                )));
+            }
+
+            match &selfi.coin_type {
+                EthCoinType::Eth => {
+                    if tx_from_rpc.to != Some(expected_swap_contract_address) {
+                        return MmError::err(ValidatePaymentError::WrongPaymentTx(format!(
+                            "Payment tx {:?} was sent to wrong address, expected {:?}",
+                            tx_from_rpc, expected_swap_contract_address,
+                        )));
+                    }
+
+                    let function = SWAP_CONTRACT
+                        .function("ethPayment")
+                        .map_to_mm(|err| ValidatePaymentError::InternalError(err.to_string()))?;
+                    let decoded = function
+                        .decode_input(&tx_from_rpc.input.0)
+                        .map_to_mm(|err| ValidatePaymentError::TxDeserializationError(err.to_string()))?;
+                    if decoded[0] != Token::FixedBytes(swap_id.clone()) {
+                        return MmError::err(ValidatePaymentError::WrongPaymentTx(format!(
+                            "Invalid 'swap_id' {:?}, expected {:?}",
+                            decoded, swap_id
+                        )));
+                    }
+
+                    if decoded[1] != Token::Address(selfi.my_address) {
```

Comment:
Good catch, I overlooked this since the integration tests were working without a problem. That was due to another bug, which I also fixed. I will write a unit test for the `watcher_validate_taker_payment` method.